src/configure/auto_mozilla.rs

```diff
@@ -224,48 +224,36 @@ fn parse_xml_with_address(in_emailaddr: &str, xml_raw: &str) -> Result<MozAutoco
 }
 
 /// Parses XML into `LoginParam` structure.
-fn parse_loginparam(in_emailaddr: &str, xml_raw: &str) -> Result<LoginParam, Error> {
+fn parse_loginparam(in_emailaddr: &str, xml_raw: &str) -> Result<Vec<ServerParams>, Error> {
     let moz_ac = parse_xml_with_address(in_emailaddr, xml_raw)?;
 
-    let mut login_param = LoginParam::new();
-    if let Some(imap_server) = moz_ac
+    let res = moz_ac
         .incoming_servers
         .into_iter()
-        .find(|incoming_server| incoming_server.typ == "imap")
-    {
-        login_param.imap.server = imap_server.hostname;
-        login_param.imap.port = imap_server.port;
-        login_param.imap.security = imap_server.sockettype;
-        login_param.imap.user = imap_server.username;
-    }
-
-    if let Some(smtp_server) = moz_ac
-        .outgoing_servers
-        .into_iter()
-        .find(|outgoing_server| outgoing_server.typ == "smtp")
-    {
-        login_param.smtp.server = smtp_server.hostname;
-        login_param.smtp.port = smtp_server.port;
-        login_param.smtp.security = smtp_server.sockettype;
-        login_param.smtp.user = smtp_server.username;
-    }
-
-    if login_param.imap.server.is_empty()
-        || login_param.imap.port == 0
-        || login_param.smtp.server.is_empty()
-        || login_param.smtp.port == 0
-    {
-        Err(Error::IncompleteAutoconfig(login_param))
-    } else {
-        Ok(login_param)
-    }
+        .chain(moz_ac.outgoing_servers.into_iter())
+        .filter_map(|incoming_server| {
+            let protocol = match incoming_server.typ.as_ref() {
+                "imap" => Some(Protocol::IMAP),
```

Comment:
would be nicer as a fromstr implementation

Reply:
This type comes from `provider` module. Adding `FromStr` implementation there may result in someone adding there some values that are not supported in mozilla autoconfig or spelled differently in mozilla and outlook autoconfig. This parsing code is mozilla-specific. In this case, mozilla needs its own `Protocol` type, which is converted into another `Protocol` type later.

Reply:
I see